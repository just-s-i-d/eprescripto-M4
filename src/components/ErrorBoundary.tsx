import { Button, Spin } from "antd";
import { Component, ReactNode, Suspense } from "react";

interface ErrorBoundaryProps {
  children: ReactNode;
  refreshComponent?: VoidFunction;
  error?: boolean;
}

interface ErrorBoundaryState {
  hasError: boolean;
}
class ErrorBoundary extends Component<ErrorBoundaryProps, ErrorBoundaryState> {
  constructor(props: ErrorBoundaryProps) {
    super(props);
    this.state = { hasError: false };
  }
  componentDidCatch(error: Error): void {
    console.trace(error); //eslint-disable-line
    this.setState({ hasError: true });
  }
  render() {
    if (this.state.hasError || this.props.error) {
      return (
        <div className="flex flex-col gap-1 justify-center items-center">
          <h2 className="text-lg font-bold">Something went wrong</h2>
          {this.props.refreshComponent && (
            <Button onClick={this.props.refreshComponent}>Try again</Button>
          )}
        </div>
      );
    }
    return <Suspense fallback={<Spin />}>{this.props.children}</Suspense>;
  }
}

export default ErrorBoundary;
