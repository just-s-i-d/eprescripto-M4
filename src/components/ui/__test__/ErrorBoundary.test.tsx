import { screen, render, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import ErrorBoundary from "../../ErrorBoundary";

describe("Error Boundary tests", () => {
  it("Renders children of Error boundary when there is no error", () => {
    const { asFragment } = render(
      <ErrorBoundary error={true}>Test Content</ErrorBoundary>,
    );
    render(<ErrorBoundary error={false}>Test Content</ErrorBoundary>);
    expect(screen.getByText("Test Content")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("render error boundary for error with try again button", () => {
    const mockedFunction = jest.fn();
    render(
      <ErrorBoundary error={true} refreshComponent={mockedFunction}>
        Test Error
      </ErrorBoundary>,
    );
    expect(screen.queryByText("Test Error")).not.toBeInTheDocument();
    expect(screen.getByText("Something went wrong")).toBeInTheDocument();
    const tryAgainBtn = screen.getByRole("button", { name: "Try again" });
    expect(tryAgainBtn).toBeInTheDocument();
    fireEvent.click(tryAgainBtn);
    expect(mockedFunction).toHaveBeenCalled();
  });
});
