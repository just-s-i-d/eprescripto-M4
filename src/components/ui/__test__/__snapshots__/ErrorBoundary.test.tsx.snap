// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error Boundary tests Renders children of Error boundary when there is no error 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col gap-1 justify-center items-center"
  >
    <h2
      class="text-lg font-bold"
    >
      Something went wrong
    </h2>
  </div>
</DocumentFragment>
`;
